File: src/c/quick_select.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define N 100
#define K 10

static int seq[N];

void init();
void print_seq();
int partition(int lo, int hi);
void swap(int p, int q);
int select(int k);

int main()
{
    init(); 
    printf("%d\n", select(K)); 
    return 0;
}

void init()
{
    for(int i = 0; i < N; i++)
        seq[i] = i;

    srand(time(NULL));

    for(int i = 1; i < N; i++)
    {
        int rnd = rand() % i;
        swap(i, rnd);
    }

    print_seq();
}

int select(int k)
{ 
    int lo = 0, hi = N - 1;
    while(hi > lo)
    {
        int j = partition(lo, hi);
        if(j < k) 
            lo = j + 1;
        else if (j > k) 
            hi = j - 1;
        else 
            return seq[k];
    }
    return seq[k];
}

int partition(int lo, int hi)
{ 
    int i = lo, j = hi + 1;

    while(1)
    {
        while(seq[++i] < seq[lo])
            if(i == hi) break;
        while(seq[--j] > seq[lo])
            if(j == lo) break;

        if(i >= j) break;
        swap(i, j);
    }
    swap(lo, j);
    return j;
}

void swap(int p, int q)
{
    int t = seq[p];
    seq[p] = seq[q];
    seq[q] = t;
}

void print_seq()
{
    for(int i = 0; i < N; i++)
        printf("%d ", seq[i]); 
    printf("\n"); 
}

File: src/cs/quick_select.cs
using System;

public class Program
{
    private static int[] seq = new int[10];

	public static void Main()
    { 
        Init();
        Console.WriteLine(Select(5));    
    }

    private static int Select(int k)
    {
        int lo = 0, hi = seq.Length - 1;
        while(true)
        {
            if(lo >= hi)
                break;
            int j = Partition(lo, hi);
            if(j > k)
                hi = j - 1;
            else if(j < k)
                lo = j + 1;
            else
                return seq[k];
        }

        return seq[k];
    }

    private static int Partition(int lo, int hi)
    {
        int i = lo, j = hi + 1;
        while(true)
        {
            while(seq[++i] < seq[lo])
                if(i == hi) break;
            while(seq[--j] > seq[lo])
                if(j == lo) break;

            if(i >= j) break;
            Swap(i, j);
        }
        Swap(lo, j); 
        return j; 
    }

	private static void Init()
    {
        for(int i = 0; i < seq.Length; i++)
            seq[i] = i;

        Random rnd = new Random();
        for(int i = 1; i < seq.Length; i++)
        {
            int r = rnd.Next(i);
            Swap(r, i); 
        }
    }

	private static void PrintSeq()
    {
        for(int i = 0; i < seq.Length; i++)
            Console.Write(seq[i] + " ");    
        Console.WriteLine();
    }


	private static void Swap(int p, int q)
    {
        int t = seq[p];
        seq[p] = seq[q];
        seq[q] = t;
    }
}

